### ISSUE TITLE
Setting a Attribute without Value

### ISSUE THREAD

**Issue by jogibear9988** [AngleSharp/AngleSharp#434]
I have a Attribute without a Value wich I'd like to set on a Element: 

```
 element.SetAttribute("readonly", null);
```

how can I do this?

**Comment by FlorianRappl** [comment]
I would consider this a bug. Intentionally, it was designed to not take `null` values. But this is wrong. So attributes without a value are not working in AngleSharp. I scheduled it for v0.10.0, but maybe I include it in a HF for v0.9.9.

**Comment by mausworks** [comment]
@FlorianRappl side note:

If the functionality is to match the JS-behavior, then passing `null` into `setAttribute` should procude `attribute-name="null"`

Passing an empty string procudes the expected result.

See my pen: https://codepen.io/MausWorks/pen/QKxBoW

**Comment by FlorianRappl** [comment]
So thinking about it for the second time (thanks at @mausworks to point that out) I think it is not directly an issue. Remark: I am not trying to mimick JS behavior - that behavior comes from the implied cast (`any` -> `string` in the JS to native transition).

Every value requires a string. If no value is given (from the HTML source) then its the empty string. Now we can talk about the serialization in such cases, but unless you show me a difference from AngleSharp vs., e.g., Chrome, in that regard I see no issue.

**Comment by FlorianRappl** [comment]
So answer to the question: There is no attribute "without" a value. Every attribute has a value.

``` cs
element.SetAttribute("readonly", String.Empty);
```

is what you actually look for (at least that's my opinion). The `readonly` (as an example) only cares about whether its available or not - not what value it has.

**Comment by mausworks** [comment]
I think the only argument for having an attribute "without" a value, is that it looks prettier in markup.

`<input type="text" disabled />` has the exact same behavior as `<input type="text" disabled="" />`. and as far as the DOM is concerned, the value of the `disabled`-attribute is still an empty string.

I'm aware that the `"null"` comes from the `String(null)` call. I just wasn't sure how closely AngleSharp would mimic the behavior. But as calling `setAttribute("foo", null)` is perfectly valid in JS; one might expect this behavior to work in AngleSharp as well, and if so; one might expect the same result.

**Comment by FlorianRappl** [comment]
Yeah, as I said: The behavior comes from the JS to native transition and the cast to a string. We do not have the JavaScript types and our instance is already a (null) string. So writing "null" would not make sense to me. If you plug in a scripting engine then yes - you get the same behavior as expected from JS.

Similarly, using some arbitrary object in JavaScript will result in its string representation (e.g., `[object Object]` for a general one), but in C# you'll get a type mismatch exception (if done via reflection; otherwise a compilation error occurs of course). So the exact JS behavior cannot be matched anyway.

Long story short: My aim is not to copy the behavior tied to JS, but the right behavior (i.e., the behavior as defined from the W3C - which is independent of the used language).

Back to the topic: I agree - this is mostly about aesthetics and the serialization of the DOM. So I would say everything is as it should be.

**Comment by zaher-aa** [comment]
> I have a Attribute without a Value wich I'd like to set on a Element:
> 
> ```
>  element.SetAttribute("readonly", null);
> ```
> 
> how can I do this?

You can use `element.toggleAttribute("readonly")`.
So if the attribute 'readonly' does exist in the element, it'll be added, otherwise, it'll be removed.
